### PR TITLE
[ramda] feat: experimental overload selection generic parameter

### PR DESCRIPTION
**Pull request by somebody1234**
This is very low priority, mostly just trying to figure out whether this is even a good idea
The new typings (mostly) shouldn't be a breaking change, but I'm not sure whether they're even useful

For context, I was meaning to deprecate `Pred` next... but was sidetracked. There are much more important things to do so feel free to just ignore this one or mark it as a draft or close it or something

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: N/A

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## ramda ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ramda@latest/))
was missing the following properties:
1. collectBy
1. modify
1. modifyPath
1. partialObject
1. on

<details>
<summary>as well as these 5 other properties...</summary>
<p>promap, sequence, splitWhenever, unwind, whereAny</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 8f5be427b9355a7b239969d8e721638a55771033
</p>

File: types/ramda/index.d.ts [hunk 1]
```diff
@@ -2069,6 +2077,9 @@ export function pipe<TArgs extends any[], R1, R2>(
     f1: (...args: TArgs) => R1,
     f2: (a: R1) => R2,
 ): (...args: TArgs) => R2;
+// For inference, so should be first. Works with generics and overloads
+// Useful for seeing which overload is picked by default
+export function pipe<F extends (...args: any[]) => unknown>(f1: F): F;
```
Comment (adispring) [inline types/ramda/index.d.ts:2082]: I don't think we should add this overload, this will swallow the above overloads' ts errors.

**Review by adispring — CHANGES_REQUESTED** [review]

File: types/ramda/index.d.ts (hunk 1 above)
Comment (adispring) [reply types/ramda/index.d.ts:2082]: Maybe we can rewrite `pipe`'s type in one more elegant way。

Comment (somebody1234) [reply types/ramda/index.d.ts:2082]: Maybe. Are you sure they will swallow the above overloads' errors? This one only takes one argument, right? So the only error it should swallow is the one for the overload below.

If you mean that it swallows it because it's below all the others (not sure about this one), I could try just moving it to the top of the overload list

File: types/ramda/index.d.ts [hunk 2]
```diff
@@ -173,7 +173,11 @@ export function allPass<T, TF1 extends T, TF2 extends T, TF3 extends T, TF4 exte
         PredTypeguard<T, TF6>,
     ],
 ): PredTypeguard<T, TF1 & TF2 & TF3 & TF4 & TF5 & TF6>;
-export function allPass<F extends Pred>(preds: readonly F[]): F;
+// For inference, so should be first. Works with generics and overloads
+export function allPass<Pred extends (...args: any[]) => unknown, _Overload extends 'pred' = 'pred'>(preds: readonly Pred[]): Pred;
```
Comment (essenmitsosse) [inline types/ramda/index.d.ts:177]: What is `_Overload` doing, since it isn't used?

Comment (somebody1234) [reply types/ramda/index.d.ts:177]: Check the tests - it's added so that you can select the overload. In this case it's not useful at all, it's just on this specific overload for consistency

Really, the only real usecase is if you want to pass a tuple as a single arg - it would normally be caught by the `args` overload, but if you manually specify `'arg'` you'd get the desired behavior.

I do think it's not very useful for `allPass` and `anyPass` specifically so it's still fine to reject this PR
I think there are some other places where it *would* be potentially useful - seems like in the overhaul PR I added it so you would be able to select which of the array/record/functor overloads to use, since they all take the same type arguments (input and output)

**Comment by somebody1234** [comment]
Note that I considered prefix `_Overload` (in the overhaul PR) - however I think having to duplicate every overload (or most) is worth the improvement in how it reads (`anyPass<'args'` could almost be read as `anyPass.args<T>()` or similar)

File: types/ramda/index.d.ts (hunk 2 above)
Comment (essenmitsosse) [reply types/ramda/index.d.ts:177]: My main concern with this is, that adding Generics like that to "select" an overload is not really documented TypeScript behaviour. So even if it would be useful, very few people would use it, because no one would know its there. Usually I don't want to look at the actual type of a function to work with it, which would be necessary to discover this "feature".

I also think its potentially confusing for maintainers as well.

Or did I misunderstand something here?

Comment (somebody1234) [reply types/ramda/index.d.ts:177]: That's a very good point, and yeah... I don't know. I did mention in the initial message that I'm mostly just trying to figure out whether this is a good idea in the first place.

Indeed it isn't documented TypeScript behaviour, but worth noting that ts-toolbelt uses something similar. Not that ts-toolbelt should be considered a good reference...

I guess what it boils down to is, is there any situation in which adding this would be useful?
I'd say... *maybe*, I don't really know - there are a ton of overloads, some of which have the same argument lists

I guess I forgot to mention, here's the implementation that was in the overhaul PR:
```typescript
export function map<_Overload extends 'array', T, U>(fn: (x: T) => U): (list: readonly T[]) => U[];
export function map<_Overload extends 'record', T, U>(fn: (x: T) => U): <K extends string>(list: Record<K, T>) => Record<K, U>;
export function map<_Overload extends 'functor', T, U>(fn: (x: T) => U): (functor: Functor<T>) => Functor<U>;
```
While I'm not 100% sure, I suspect this would come in useful when passing `map` into any of the higher order functions, otherwise IIRC `map` can only work for one type - currently that is arrays, since they're the bottom-most overload in the list.
In other words, it's currently impossible to specify "I want to use `.map(fn): (functor: Functor<T>) => Functor<U>`" when you're passing it to say, `pipe`, or `converge`

Comment (essenmitsosse) [reply types/ramda/index.d.ts:177]: I think TypeScript is totally capable of typing this. Didn't test it, but something like this should be fine, and type inference in `pipe` should work (if `pipe` is typed correctly, which I actually didn't check):

```typescript
export function map<T, U>(fn: (x: T) => U): {
    (list: readonly T[]): U[],
    <K extends string>(list: Record<K, T>): Record<K, U>,
    (functor: Functor<T>): Functor<U>
}
```
[Playground](https://www.typescriptlang.org/play?ssl=5&ssc=2&pln=1&pc=1#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwLYCGYAPACoA0cAqgHwAUCKAXHHSC6QJRwC8N1nFgG8AUHHGsANpgDOMFlGAEAJjkkBPOKQDaAXUHU95MROIBpOKBjAUymXDlRMKAOb1pclgCVgaaMvNKUhoDHz8oALNKWmMJVmR0GGgWADFUDGgyENT0pKhiWhEAXxERPxQ5fCRJGAAmXnwiOjoCFhQkPAAjYChuPjgCADokgGUYJ1c6Tk4AblLyyus5AEEoKAJNHiqa2rptAEZyWvIAZn0ynEXgOTD-BrxqurohVrhDuE6WY7g0FhOizhAA)

Comment (essenmitsosse) [reply types/ramda/index.d.ts:177]: Also I really like the structure this gives, because it groups overloads, that start out the same.

Comment (somebody1234) [reply types/ramda/index.d.ts:177]: I do believe `pipe` does, in fact, not work:
[Playground](https://www.typescriptlang.org/play?ssl=15&ssc=9&pln=14&pc=1#code/JYWwDg9gTgLgBAKjgQwM5wEpwGZQiOAIimRABNlCBuAKBoFMAPSWHAVwDsBjGYCDuCGRgAPABUANHACqAPgAU2DgC448xqrEBKOAF5ZMraoDeNOObUAbYKhiqo9ZGX6WAnnDEBtALpGZPiTMLEQBpOCYYeg4ydFsoYA4AcwVrW1UMei5oMlCpMVk-DKyoHJCpOUCLNWxOHmhVADFamGhxAsbm1rkaAF86LI5bQTZLGAAmPUFheXlkVQ42EAAjeigdfRQAOhaAZRh4pPktLVoaAaHI2wBBKBJ3XWHRsflPAEYJMYkAZl8z-gv6LYitlJiARuN5MY5nB3nAlqpPnAuKovj0tP1-vAIAALDgQSYYTZgYBgejyIRgGbQhbLVbrAzIbYQPYHRJHY60AD0nKqAD0APxAA)

In this PR I've changed the type of `pipe` to work nicely for one parameter, but it still will not work for more than one parameter.

Comment (essenmitsosse) [reply types/ramda/index.d.ts:177]: At work we actually use our own version of `pipe` which works great. I'll take a look at Ramda `pipe` the other day, because its definitely typeable.

Comment (somebody1234) [reply types/ramda/index.d.ts:177]: re: the overload grouping - that still exists. In the case of `map`, those are extra overloads.
Agree that the overload returning an overloaded function should exist and is a huge boon to usability when `map` is used at the top level

**Comment by somebody1234** [comment]
closing for now, very janky. maybe worth searching for alternatives, i still do think this is a problem for `map` overloads
